src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs

```diff
@@ -135,14 +135,18 @@ public bool IsEnabled(LogLevel logLevel)
 
         private static IReadOnlyList<KeyValuePair<string, object?>>? ProcessState<TState>(LogRecord logRecord, TState state, bool parseStateValues)
         {
-            /* TODO: Enable this if/when LogRecordAttributeList becomes public.
-            if (state is LogRecordAttributeList logRecordAttributes)
+            if (typeof(TState) == typeof(LogRecordAttributeList))
```

Comment:
I know we've discussed before, but remind me of a use case where this will be the case...

I understand end users generally do not use the low-level [Log method](https://learn.microsoft.com/dotnet/api/microsoft.extensions.logging.ilogger.log) accepting a `TState` type param. So is it that we anticipate extension authors may use it and also use OpenTelemetry's `LogRecordAttributeList`?

Reply:
Not sure yet. The way I implemented it, the JIT should basically remove this code if it isn't used. So I figured... no harm in including the path. What this change does is allow people to use it if they are interested in doing so. I'm currently working on complex type logging. What I'm going to explore is an "extensions" type of package that includes helpers and possibly source generator(s) to do that in a high-perf way. We could also do the same for this path.

Why would anyone want to use this path?

* [LoggerExtensions](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerExtensions.cs) use `params object?[] args` so they always allocate an array and box things up. TState ends up [FormattedLogValues](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/FormattedLogValues.cs). That struct is going to get [boxed here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/a6c3cac15d20cb39545c0a564ca0595fada607ee/src/OpenTelemetry/Logs/OpenTelemetryLogger.cs#L103). It may also be copied if batching is in play. So lots of perf penalties to using those extensions.

* The [logging source generators](https://learn.microsoft.com/en-us/dotnet/core/extensions/high-performance-logging) are better because they don't allocate an array you get a nice purpose-built struct as TState. But it will still get boxed in the same way and possibly copied if batching is in play.

The advantage to using this path for `LogRecordAttributeList` is we avoid a lot of that boxing and we copy the stack values to the pool storage directly so there isn't any additional buffering that needs to happen.

The signature for all scenarios is still `IReadOnlyList<>` of `KeyValuePair<string, object>` items so a lot of value boxing goes on (int tag for example) but not much I can do about that 😢